### ISSUE TITLE
Interrupting flow

### ISSUE THREAD

**Issue by SaracenOne** [kagenash1/godot-behavior-tree#6]
This may already be covered, but the current implementation doesn't provide any obvious solution on how to do this. If you an AI agent perform a sequence using things like wait timers or other yields, there doesn't seem to be any way to get them to re-evaluate from a higher level node based on external stimuli. An example might be an AI who simply wanders around, but if they get attacked,  you would want their wander sequence interrupted and may instead go into an attack sequence.

**Comment by GabrieleTorini** [comment]
I see what you mean. 
If you want to make a branch conditional, you put it under a decorator, extend the decorator and define your condition there (as it's done in the example), but if your sequence (or any other node) is in a running state it can't be interrupted until completion. 

I will definitely work on making this possible, maybe through with an event driven approach as it's done in Unreal. Thank you for bringing it up, this is a work in progress and I'm trying to improve it everyday!

**Comment by SaracenOne** [comment]
Yeah, trying parallel nodes, I was able to develop some AI which addressed my problem reasonably well. I think taking your advice of thinking about behaviour trees less in the terms of finte state machines helps too. I am fairly new to them. That said, I still think there's value in the Unreal-style approach of branch aborting, especially as trees start to balloon in complexity.

**Comment by GabrieleTorini** [comment]
Good :) Parallels are the "worst" in terms of performance so, if you can, think of other ways. 
Always happy to provide support if you need! I know it's not very user friendly yet but I'm actively trying to improve it and make it as complete and powerful as possible. 
I'm onto the flow abortion now so it will be a thing in the next update. Also, the way running states are handled will change with it, so I'm gonna have to make sure the update doesn't break your behavior trees.

**Comment by MadFlyFish** [comment]
Thanks for your great addons! I have being using it for a couple of weeks, and still looking forward to more  future updates.
Recently I am developing a tool, which uses Xmind and excel as an tree editor and create behavior tree dynamically in run-time, mostly based on your addons.
That helps my work a lot. Therefore I am introducing your addons to my Godot developer friends in China, and they show great interest too.
 May I have your email or other contacts and learn some more about AI? Thanks.

![image](https://user-images.githubusercontent.com/36298330/111722049-0a0f9080-889c-11eb-865e-ea2bdc47456e.png)
![image](https://user-images.githubusercontent.com/36298330/111722079-185dac80-889c-11eb-96f1-5506ccdb0395.png)
![image](https://user-images.githubusercontent.com/36298330/111722099-21e71480-889c-11eb-84cc-3551c5e0d12f.png)
![image](https://user-images.githubusercontent.com/36298330/111722447-d5500900-889c-11eb-9e59-577e3ab2c591.png)
![image](https://user-images.githubusercontent.com/36298330/111722489-eac53300-889c-11eb-8b9b-3f6c0cd0019f.png)
![image](https://user-images.githubusercontent.com/36298330/111722509-f153aa80-889c-11eb-9797-d59abd682084.png)

**Comment by GabrieleTorini** [comment]
It's amazing to see this! I'm glad my addon turned out useful and I'm excited to see what you've done with it. 
My email is gabriele.torini1@gmail.com or if you want you can find me on Discord at kagenashi#8224

**Comment by GabrieleTorini** [comment]
This issue will be addressed in the 1.2.rc4 version.

**Comment by SaracenOne** [comment]
Nice!

**Comment by GabrieleTorini** [comment]
I have started updating the 1.2.rc4 branch. There are some changes already, although the features is not there yet. However, I made a much more complex example to show how to achieve some of the behaviors you people were talking about!

**Comment by mcccclean** [comment]
Hello, weighing in on this quite late - just here to say that I've had some success implementing this locally without too much refactoring, and I wanted to share my approach in case it is useful toward the official implementation (or to others finding this thread).

In bt_node.gd (unrelated sections elided):

```gdscript
signal completed(result)
const ABORTED = -1;

func abort():
  fail()
  complete(ABORTED)

func complete(value):
  if running():
    return
  emit_signal("completed", value)

func tick(agent: Node, blackboard: Blackboard) -> bool:
  # .... etc ...
  if result is GDScriptFunctionState:
    assert(running(), "BTNode execution was suspended but it's not running. Did you succeed() or fail() before yield?")
    result.connect("completed", self, "complete")
    result = yield(self, "completed")
    if result is int and result == ABORTED:
	    return
```

and then in BehaviourTree:

```gdscript
# halt the behaviour tree
func interrupt() -> void:
  bt_root.propagate_call("abort")

# halt the behaviour tree and restart it
# (useful for scenarios where an agent is surprised, eg taking damage, enemy arrives in range etc)
func hard_restart() -> void:
  interrupt()
  is_active = true
  start()
```

The central change there is that the `yield` inside `tick` can now be terminated early by calling `complete` on the node, rather than always having to wait for the underlying function call to finish. This allows for mechanisms that short circuit the node into a success or failure state as well as aborting it.

Of course there are some downsides to this approach - the function that the node was waiting on will still continue its execution uninterrupted, which might have strange side effects. This won't come up if nodes are designed to only do one thing (like, a node whose job is to wait and then execute a side effect should probably be split up into a waiter and an executor node) but perhaps this is too big of a tripping hazard to include without further safeguards. I also don't know how performant it is; my project only includes a handful of actors so it might have additional problems when dozens/hundreds of agents are in the scene.

Thanks @GabrieleTorini for the great module - it's so useful and extensible, it's been a great help to my hobby projects. I hope this is useful to you!

**Comment by GabrieleTorini** [comment]
Thanks! This is definitely a great addition.
Would you like to make a PR for this? I'll try to see if it works with the sample project.
(I'm not a geet (git geek) so the whole process of the PR is probably gonna be clunky)

I'm glad you're happy with the plugin. It was a big experiment, and thinking back I think it could be done in an entirely different way, but as long as it's stable and gets you to make your AI, I'm satisfied!

**Comment by ATHellboy** [comment]
> Hello, weighing in on this quite late - just here to say that I've had some success implementing this locally without too much refactoring, and I wanted to share my approach in case it is useful toward the official implementation (or to others finding this thread).
> 
> In bt_node.gd (unrelated sections elided):
> 
> ```gdscript
> signal completed(result)
> const ABORTED = -1;
> 
> func abort():
>   fail()
>   complete(ABORTED)
> 
> func complete(value):
>   if running():
>     return
>   emit_signal("completed", value)
> 
> func tick(agent: Node, blackboard: Blackboard) -> bool:
>   # .... etc ...
>   if result is GDScriptFunctionState:
>     assert(running(), "BTNode execution was suspended but it's not running. Did you succeed() or fail() before yield?")
>     result.connect("completed", self, "complete")
>     result = yield(self, "completed")
>     if result is int and result == ABORTED:
> 	    return
> ```
> 
> and then in BehaviourTree:
> 
> ```gdscript
> # halt the behaviour tree
> func interrupt() -> void:
>   bt_root.propagate_call("abort")
> 
> # halt the behaviour tree and restart it
> # (useful for scenarios where an agent is surprised, eg taking damage, enemy arrives in range etc)
> func hard_restart() -> void:
>   interrupt()
>   is_active = true
>   start()
> ```
> 
> The central change there is that the `yield` inside `tick` can now be terminated early by calling `complete` on the node, rather than always having to wait for the underlying function call to finish. This allows for mechanisms that short circuit the node into a success or failure state as well as aborting it.
> 
> Of course there are some downsides to this approach - the function that the node was waiting on will still continue its execution uninterrupted, which might have strange side effects. This won't come up if nodes are designed to only do one thing (like, a node whose job is to wait and then execute a side effect should probably be split up into a waiter and an executor node) but perhaps this is too big of a tripping hazard to include without further safeguards. I also don't know how performant it is; my project only includes a handful of actors so it might have additional problems when dozens/hundreds of agents are in the scene.
> 
> Thanks @GabrieleTorini for the great module - it's so useful and extensible, it's been a great help to my hobby projects. I hope this is useful to you!

Hey there,

I'm using this piece of code for hard restarting the bt but it doesn't work for me. At least it doesn't work when the current leaf is under RepeatUntil. Have you checked it with RepeatUntil too ?
